### PR TITLE
[Bugfix] Fix empty tensors may being treated as pinned

### LINKED ISSUE #5004: [Bug] `edge_subgraph` would fail with UVA when there're empty relations
## 🐛 Bug

`edge_subgraph`/`khop_in_subgraph`/`khop_out_subgraph` would fail with UVA when there're empty relations in the graph.

## To Reproduce

```python
import dgl
import torch

def create_test_heterograph2(idtype):
    """test heterograph from the docstring, with an empty relation"""
    # 3 users, 2 games, 2 developers
    # metagraph:
    #    ('user', 'follows', 'user'),
    #    ('user', 'plays', 'game'),
    #    ('user', 'wishes', 'game'),
    #    ('developer', 'develops', 'game')

    g = dgl.heterograph(
        {
            ("user", "follows", "user"): ([0, 1], [1, 2]),
            ("user", "plays", "game"): ([0, 1, 2, 1], [0, 0, 1, 1]),
            ("user", "wishes", "game"): ([0, 2], [1, 0]),
            ("developer", "develops", "game"): ([], []),
        },
        idtype=idtype,
    )
    return g

def test_uva_subgraph(idtype):
    device = torch.device("cuda:0")
    g = create_test_heterograph2(idtype)
    g.create_formats_()
    g.pin_memory_()
    indices = {"user": torch.tensor([0], dtype=idtype, device=device)}
    edge_indices = {"follows": torch.tensor([0], dtype=idtype, device=device)}

    # all the following three tests would fail
    assert g.edge_subgraph(edge_indices).device == device
    assert g.khop_in_subgraph(indices, 1)[0].device == device
    assert g.khop_out_subgraph(indices, 1)[0].device == device
    g.unpin_memory_()

test_uva_subgraph(torch.int64)
```

## Expected behavior

They should work.

## Environment

 - DGL Version (e.g., 1.0): master branch
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3):
 - OS (e.g., Linux):
 - How you installed DGL (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version (if applicable):
 - GPU models and configuration (e.g. V100):
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by yaox12**
## Description

Fix #5004.

The root causes are two aspects:
1. Empty tensors might be treated as pinned though they're not.
2. When an empty CSR matrix is pinned, it has the `is_pinned=true` flag, but the underlying `indices` could be not pinned (empty). We should avoid calling `cudaHostGetDevicePointer` on it.

This PR fixes the above two issues and enhances the unit tests for them.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Review by nv-dlasalle — APPROVED** [review]

File: src/array/cuda/array_index_select.cu [hunk 1]
```diff
@@ -33,9 +27,13 @@ NDArray IndexSelect(NDArray array, IdArray index) {
 
   // use index->ctx for pinned array
   NDArray ret = NDArray::Empty(shape, array->dtype, index->ctx);
-  if (len == 0) return ret;
+  if (len == 0 || arr_len == 0) return ret;
```
Comment (chang-l) [inline src/array/cuda/array_index_select.cu:30]: Just wondering if this checking condition is complete or not... It seems only checking the length of the first dim, right? Could it cover the case when `array->shape[i] == 0`?

File: src/array/cuda/csr_get_data.cu [hunk 2]
```diff
@@ -43,19 +43,13 @@ NDArray CSRGetData(
...
-          cudaHostGetDevicePointer(&data_data, csr.data.Ptr<IdType>(), 0));
-    }
-  }
+  const IdType* indptr_data =
+      static_cast<IdType*>(cuda::GetDevicePointer(csr.indptr));
+  const IdType* indices_data =
+      static_cast<IdType*>(cuda::GetDevicePointer(csr.indices));
+  const IdType* data_data =
+      CSRHasData(csr) ? static_cast<IdType*>(cuda::GetDevicePointer(csr.data))
+                      : nullptr;
```
[the remark this answers is not in the document]
Comment (yaox12) [reply src/array/cuda/csr_get_data.cu:52]: We will check if `csr.indptr/indices/data` is pinned separately in this `GetDevicePointer()` function. A CSR matrix could hold the `is_pinned=true` flag even if it's empty, so we should avoid calling `cudaHostGetDevicePointer` on it's `indptr/indices/data` in such a case.
https://github.com/dmlc/dgl/blob/e28f07812bae274f5ba49d26629de5541c9cff60/include/dgl/aten/csr.h#L137-L145

File: src/array/cuda/array_index_select.cu (hunk 1 above)
Comment (yaox12) [reply src/array/cuda/array_index_select.cu:30]: I cannot imagine a tensor has other dims when the first dim is 0.

Comment (yaox12) [reply src/array/cuda/array_index_select.cu:30]: Here is the utility to check a null array.
https://github.com/dmlc/dgl/blob/e28f07812bae274f5ba49d26629de5541c9cff60/include/dgl/aten/array_ops.h#L37

File: include/dgl/aten/macro.h [hunk 3]
```diff
@@ -309,7 +309,9 @@
   });
 
 #define CHECK_VALID_CONTEXT(VAR1, VAR2)                          \
-  CHECK(((VAR1)->ctx == (VAR2)->ctx) || (VAR1).IsPinned())       \
+  CHECK(                                                         \
+      ((VAR1)->ctx == (VAR2)->ctx) || (VAR1).IsPinned() ||       \
+      ((VAR1).NumElements() == 0))                               \
```
Comment (chang-l) [inline include/dgl/aten/macro.h:314]: Do you think if it would be possible or better to check 0-element and return early [here](https://github.com/dmlc/dgl/blob/e28f07812bae274f5ba49d26629de5541c9cff60/src/array/array.cc#L120)?

**Review by chang-l — APPROVED** [review]

File: src/array/cuda/array_index_select.cu (hunk 1 above)
Comment (chang-l) [reply src/array/cuda/array_index_select.cu:30]: I see...

Comment (yaox12) [reply src/array/cuda/array_index_select.cu:30]: I get your point. It would be better to use `array.NumElements()` here.

Comment (yaox12) [reply src/array/cuda/array_index_select.cu:30]: Or simply `arr_len * num_feat` here.

File: include/dgl/aten/macro.h (hunk 3 above)
Comment (yaox12) [reply include/dgl/aten/macro.h:314]: I agree. But I find we need to return early for many other functions, such as `dgl::aten::CSRSliceRows()`...

I will add a comment to this macro.
